### DISCUSSION TITLE
Subresource Integrity

### DISCUSSION THREAD

**Discussion by robsonsobral** [ExpressionEngine/ExpressionEngine#3390] [Feature Requests]
Hi!

Can we, please, have file hashes to allow the use of `integrity` attribute for [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)?

- [x] I am capable and would like to work on implementation of this feature if it is considered.

**Comment by intoeetive** [comment]
@robsonsobral can you give an example of how this might look like?

**Comment by robsonsobral** [comment]
Oh, maybe I should have opened this issue on the docs. I'm sorry. It is so obvious to me that I didn't think of that.

```html
{news_image}
  <img
    src="{url}"
    alt="{description}"
    width="{width}"
    height="{height}"
    crossorigin="anonymous"
    integrity="{fingerprint hash="sha512"}"
  />
{/news_image}
```

The default value could be a SHA256 saved at the database. All options:

- `sha256`
- `sha384`
- `sha512`
- `md5`

Example of parsed `{fingerprint hash="sha384"}` with base64-encoded sha384 hash:

```
sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC
```

Being `oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC` the "hash" part, and the prefix `sha384` indicates that it's a sha384 hash.

To [generate the hash, use](https://www.srihash.org/):

```shell
openssl dgst -sha384 -binary FILENAME.js | openssl base64 -A 
```

Scripts and stylesheets can also have `integrity` attribute. This is from a project of mine:

```html
<link
  rel="stylesheet"
  href="/stylesheets/index.38bd61bc190f0c181d52ee9f06d1c8c14c13cd0a1a647cbaee505cdf8fdfca4eade18b1a04fb7af1ed7fc16ae21c889a6ac879031950799e24efc77073d16de4.css"
  integrity="sha512-OL1hvBkPDBgdUu6fBtHIwUwTzQoaZHy67lBc34/fyk6t4YsaBPt68e1/wWriHIiaash5AxlQeZ4k78dwc9Ft5A=="
  media="all"
  crossorigin="anonymous"
/>
```

Please, see #1220

**Comment by robsonsobral** [comment]
Keeping my fight to make eecms more CDN friendly, I've been thinking on this feature.

A first step can be to have a fingerprint for templates:

```html
<link integrity="{fingerprint='channel/channel_css' hash='sha512'}" rel="stylesheet" href="{stylesheet='channel/channel_css'}" />
```

The `fingerprint` tag could be part of the Resource Library added on the [PR #1259](https://github.com/ExpressionEngine/ExpressionEngine/pull/1259)

**Comment by intoeetive** [comment]
This sounds more like an add-on, but I'll keep this as core feature request as well. Converting to discussion, as that's what we use for FRs now

**Reply by robsonsobral** [reply]
I disagree on making an add-on. Security features, like SRI, CSP, CORS..., should be part of the job. However, I recognize that SRI is more a concern on 3rd-party content.
